### PR TITLE
Create MongoSinkTask unit tests for DQL-related behavior

### PR DESCRIPTION
**Pull request by stIncMale**
This is a subtask of KAFKA-257 created to produce smaller PRs. It is based on the changes done in https://github.com/mongodb/mongo-kafka/pull/88.

KAFKA-268

### CODE REVIEW

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java [hunk 1]
```diff
@@ -291,4 +226,91 @@ private String generateWriteErrors(
     }
     return "[" + String.join(", ", errorString) + "]";
   }
+
+  static final class StartedMongoSinkTask {
```
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java:230]: The refactoring in `MongoSinkTask` did not change any logic, but merely extracted the instance logic that is not related to the Kafka Connect-specific lifecycle of the `MongoSinkTask` to a new class `StartedMongoSinkTask`, which can now be easily unit tested.

File: src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                // batch1
+                Records.simpleValid(TEST_TOPIC, 0),
+                // batch2
+                Records.simpleValid(TEST_TOPIC2, 1),
+                Records.simpleValid(TEST_TOPIC2, 2),
+                Records.simpleValid(TEST_TOPIC2, 3),
+                // batch3
+                Records.simpleValid(TEST_TOPIC, 4)),
+            asList(0, 1, 2, 3, 4),
+            emptyList());
```
Comment (stIncMale) [inline src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java:188]: KAFKA-257 will change this empty expectation to `asList(2, 3)`:
- 2 will be reported because it has a corresponding write error;
- 3 will be reported because the bulk write is ordered, which means that 3 is not attempted to be written.

**Review by rozza — APPROVED** [review]
Looks good - some stylistic comments for consideration but not necessarily requiring action.

File: src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    }
+
+    static SinkRecord simpleInvalid(final String topicName, final int idx) {
+      return new SinkRecord(
+          topicName, 0, Schema.STRING_SCHEMA, "", Schema.STRING_SCHEMA, "notJSON_" + idx, idx);
+    }
+
+    static boolean match(
+        final SinkRecord expected, final WriteModel<? extends BsonDocument> actual) {
+      ReplaceOneModel<BsonDocument> writeModel = cast(actual);
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java:238]: I added `assertTrue(actual instanceof ReplaceOneModel)` to make the expectation clear. If for some reason in the future this method encounters a write model different from `ReplaceOneModel`, I want it to fail as early as possible (it will need to be updated to support comparing that write model), rather than returning `false` and causing a failure in a different place.

File: src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+  private static final MongoNamespace DEFAULT_NAMESPACE =
+      new MongoNamespace(TEST_DATABASE, "myColl");
+
+  private Map<String, String> properties;
+  private FakeClient client;
+  private FakeErrorReporter errorReporter;
+
+  @BeforeEach
+  void setUp() {
+    properties = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java:92]: - Renamed `FakeClient` -> `BulkWritesCapturingClient`, `FakeErrorReporter` -> `InMemoryErrorReporter`.

- I don't think having these classes in `SinkTestHelper` is useful at this point. If we introduce other test classes that need either the same of very similar functionality, then extracting fake implementations for reusing may become useful.

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java (hunk 1 above)
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java]: I did not make it a top-level class initially, because I was trying to minimize the changes in `MongoSinkTask`. But I am happy to have such a change. Done.

> So this is just a mechanism for setting the mongo client?

Yes it's just manual dependency injection (DI):

- It results in dependencies of `StartedMongoSinkTask` (sink config, MongoDB client, error reporter) being expressed clearly, which improves readability;
- It allows the code that creates such tasks to control the implementation of dependencies, which decouples `StartedMongoSinkTask` from those implementations. This is crucial for unit tests, because they now can easily specify mocked implementations.

> the stages are all still linked to the Kafka lifestyle - eg: stop and put

The important part is that the two aforementioned points are still true. I poorly expressed the idea in the initial comment, but hopefully now it is clearer.

> Is there a cleaner way to achieve that for testing.

I don't think there is. DI is applied here right to its point. And of all different styles of DI, constructor injection is the cleanest one.

File: src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rozza) [reply src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java:92]: Ack

File: src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java (hunk 2 above)
Comment (rozza) [reply src/test/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTaskTest.java:191]: Ack

**Review by rozza — APPROVED** [review]
LGTM!
